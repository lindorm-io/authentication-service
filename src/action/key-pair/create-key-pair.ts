import Joi from "@hapi/joi";
import { IAuthContext } from "../../typing";
import { KeyType } from "@lindorm-io/key-pair";
import { assertAccountAdmin, generateKeyPair } from "../../support";

export interface ICreateKeyPairOptions {
  type: KeyType;
}

export interface ICreateKeyPairData {
  algorithm: string;
  keyPairId: string;
  type: string;
}

const schema = Joi.object({
  type: Joi.string().valid(KeyType.EC, KeyType.RSA).required(),
});

export const createKeyPair = (ctx: IAuthContext) => async (
  options: ICreateKeyPairOptions,
): Promise<ICreateKeyPairData> => {
  await schema.validateAsync(options);

  const { cache, logger, repository } = ctx;
  const { type } = options;

  await assertAccountAdmin(ctx)();

  const keyPair = await generateKeyPair(type);

  await repository.keyPair.create(keyPair);
  await cache.keyPair.create(keyPair);

  logger.debug("key pair created", {
    algorithm: keyPair.algorithm,
    id: keyPair.id,
    type: keyPair.type,
  });

  return {
    algorithm: keyPair.algorithm,
    keyPairId: keyPair.id,
    type: keyPair.type,
  };
};
