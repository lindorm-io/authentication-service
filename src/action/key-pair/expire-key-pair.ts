import Joi from "@hapi/joi";
import { IKoaAuthContext } from "../../typing";
import { InvalidExpiryString } from "../../error";
import { add } from "date-fns";
import { assertAccountAdmin } from "../../support";
import { stringToDurationObject, stringToMilliseconds } from "@lindorm-io/core";

export interface IExpireKeyPairOptions {
  keyPairId: string;
  expires: string;
}

const schema = Joi.object({
  keyPairId: Joi.string().guid().required(),
  expires: Joi.string().required(),
});

export const expireKeyPair = (ctx: IKoaAuthContext) => async (options: IExpireKeyPairOptions): Promise<void> => {
  await schema.validateAsync(options);

  const { cache, logger, repository } = ctx;
  const { expires, keyPairId } = options;

  await assertAccountAdmin(ctx)();

  if (stringToMilliseconds(expires) === 0) {
    throw new InvalidExpiryString(expires);
  }

  const keyPair = await repository.keyPair.find({ id: keyPairId });

  keyPair.expires = add(new Date(), stringToDurationObject(expires));

  const updated = await repository.keyPair.update(keyPair);

  try {
    await cache.keyPair.update(updated);
  } catch (err) {
    logger.error("could not update cache", err);
  }

  logger.debug("key pair updated", {
    id: keyPair.id,
    expires: keyPair.expires,
  });
};
