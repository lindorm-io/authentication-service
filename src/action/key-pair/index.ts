export * from "./create-key-pair";
export * from "./expire-key-pair";
