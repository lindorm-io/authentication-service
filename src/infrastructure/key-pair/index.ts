export * from "./KeyPairCache";
export * from "./KeyPairRepository";
