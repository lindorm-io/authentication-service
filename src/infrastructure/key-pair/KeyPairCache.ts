import { schema } from "./schema";
import { CacheBase, ICache, ICacheOptions } from "@lindorm-io/redis";
import { IKeyPair, KeyPair } from "@lindorm-io/key-pair";

export interface IKeyPairCache extends ICache<KeyPair> {
  create(entity: KeyPair): Promise<KeyPair>;
  update(entity: KeyPair): Promise<KeyPair>;
  find(id: string): Promise<KeyPair>;
  findAll(): Promise<Array<KeyPair>>;
  remove(entity: KeyPair): Promise<void>;
}

export class KeyPairCache extends CacheBase<KeyPair> implements IKeyPairCache {
  constructor(options: ICacheOptions) {
    super({
      client: options.client,
      entityName: "KeyPair",
      expiresInSeconds: options.expiresInSeconds,
      logger: options.logger,
      schema,
    });
  }

  protected createEntity(data: IKeyPair): KeyPair {
    return new KeyPair(data);
  }

  protected getEntityJSON(entity: KeyPair): IKeyPair {
    return {
      id: entity.id,
      version: entity.version,
      created: entity.created,
      updated: entity.updated,
      events: entity.events,

      algorithm: entity.algorithm,
      expires: entity.expires,
      passphrase: entity.passphrase,
      privateKey: entity.privateKey,
      publicKey: entity.publicKey,
      type: entity.type,
    };
  }

  protected getEntityKey(entity: KeyPair): string {
    return entity.id;
  }

  public async create(entity: KeyPair): Promise<KeyPair> {
    return super.create(entity);
  }

  public async update(entity: KeyPair): Promise<KeyPair> {
    return super.update(entity);
  }

  public async find(id: string): Promise<KeyPair> {
    return super.find(id);
  }

  public async findAll(): Promise<Array<KeyPair>> {
    return super.findAll();
  }

  public async remove(entity: KeyPair): Promise<void> {
    await super.remove(entity);
  }
}
