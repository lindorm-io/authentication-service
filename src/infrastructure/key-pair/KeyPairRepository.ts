import { IKeyPair, KeyPair } from "@lindorm-io/key-pair";
import { IRepository, IRepositoryOptions, RepositoryBase } from "@lindorm-io/mongo";
import { MongoCollection } from "../../enum";
import { indices } from "./indices";
import { schema } from "./schema";

export interface IKeyPairFilter {
  id?: string;
  algorithm?: string;
  type?: string;
}

export interface IKeyPairRepository extends IRepository<KeyPair> {
  create(entity: KeyPair): Promise<KeyPair>;
  update(entity: KeyPair): Promise<KeyPair>;
  find(entity: IKeyPairFilter): Promise<KeyPair>;
  findMany(filter: IKeyPairFilter): Promise<Array<KeyPair>>;
  findOrCreate(entity: IKeyPairFilter): Promise<KeyPair>;
  remove(entity: KeyPair): Promise<void>;
}

export class KeyPairRepository extends RepositoryBase<KeyPair> implements IKeyPairRepository {
  constructor(options: IRepositoryOptions) {
    super({
      collectionName: MongoCollection.KEY_PAIR,
      db: options.db,
      logger: options.logger,
      indices,
      schema,
    });
  }

  protected createEntity(data: IKeyPair): KeyPair {
    return new KeyPair(data);
  }

  protected getEntityJSON(entity: KeyPair): IKeyPair {
    return {
      id: entity.id,
      version: entity.version,
      created: entity.created,
      updated: entity.updated,
      events: entity.events,

      allowed: entity.allowed,
      algorithm: entity.algorithm,
      expires: entity.expires,
      passphrase: entity.passphrase,
      privateKey: entity.privateKey,
      publicKey: entity.publicKey,
      type: entity.type,
    };
  }

  public async create(entity: KeyPair): Promise<KeyPair> {
    return super.create(entity);
  }

  public async update(entity: KeyPair): Promise<KeyPair> {
    return super.update(entity);
  }

  public async find(filter: IKeyPairFilter): Promise<KeyPair> {
    return super.find(filter);
  }

  public async findMany(filter: IKeyPairFilter): Promise<Array<KeyPair>> {
    return super.findMany(filter);
  }

  public async remove(entity: KeyPair): Promise<void> {
    await super.remove(entity);
  }
}
