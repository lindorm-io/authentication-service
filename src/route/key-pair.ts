import { BEARER_TOKEN_MW_OPTIONS } from "../config";
import { HttpStatus } from "@lindorm-io/core";
import { IKoaAuthContext } from "../typing";
import { Router } from "@lindorm-io/koa";
import { accountMiddleware } from "../middleware";
import { bearerTokenMiddleware } from "@lindorm-io/koa-jwt";
import { createKeyPair, expireKeyPair } from "../action";
import { clientMiddleware } from "@lindorm-io/koa-client";

export const router = new Router();

router.use(clientMiddleware());
router.use(bearerTokenMiddleware(BEARER_TOKEN_MW_OPTIONS));
router.use(accountMiddleware);

router.post(
  "/",
  async (ctx: IKoaAuthContext): Promise<void> => {
    const { type } = ctx.request.body;

    ctx.body = await createKeyPair(ctx)({ type });
    ctx.status = HttpStatus.Success.CREATED;
  },
);

router.patch(
  "/:id/expire",
  async (ctx: IKoaAuthContext): Promise<void> => {
    const { expires } = ctx.request.body;

    await expireKeyPair(ctx)({
      keyPairId: ctx.params.id,
      expires,
    });

    ctx.body = {};
    ctx.status = HttpStatus.Success.NO_CONTENT;
  },
);
