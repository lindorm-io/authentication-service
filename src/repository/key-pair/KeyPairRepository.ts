import { IKeyPair, KeyPair } from "@lindorm-io/key-pair";
import { IRepository, IRepositoryOptions, RepositoryBase } from "@lindorm-io/mongo";
import { MongoCollection } from "../../enum";
import { indices } from "./KeyPairRepository.indices";
import { schema } from "./KeyPairRepository.schema";

export interface IKeyPairFilter {
  id?: string;
  algorithm?: string;
  type?: string;
}

export interface IKeyPairRepository extends IRepository<KeyPair> {
  create(keyPair: KeyPair): Promise<KeyPair>;
  update(keyPair: KeyPair): Promise<KeyPair>;
  find(keyPair: IKeyPairFilter): Promise<KeyPair>;
  findMany(filter: IKeyPairFilter): Promise<Array<KeyPair>>;
  remove(keyPair: KeyPair): Promise<void>;
}

export class KeyPairRepository extends RepositoryBase<KeyPair> implements IKeyPairRepository {
  constructor(options: IRepositoryOptions) {
    super({
      collectionName: MongoCollection.KEY_PAIR,
      db: options.db,
      logger: options.logger,
      indices,
      schema,
    });
  }

  protected createEntity(data: IKeyPair): KeyPair {
    return new KeyPair(data);
  }

  protected getEntityJSON(keyPair: KeyPair): IKeyPair {
    return {
      id: keyPair.id,
      version: keyPair.version,
      created: keyPair.created,
      updated: keyPair.updated,
      events: keyPair.events,

      algorithm: keyPair.algorithm,
      expires: keyPair.expires,
      passphrase: keyPair.passphrase,
      privateKey: keyPair.privateKey,
      publicKey: keyPair.publicKey,
      type: keyPair.type,
    };
  }

  public async create(keyPair: KeyPair): Promise<KeyPair> {
    return super.create(keyPair);
  }

  public async update(keyPair: KeyPair): Promise<KeyPair> {
    return super.update(keyPair);
  }

  public async find(filter: IKeyPairFilter): Promise<KeyPair> {
    return super.find(filter);
  }

  public async findMany(filter: IKeyPairFilter): Promise<Array<KeyPair>> {
    return super.findMany(filter);
  }

  public async remove(keyPair: KeyPair): Promise<void> {
    await super.remove(keyPair);
  }
}
