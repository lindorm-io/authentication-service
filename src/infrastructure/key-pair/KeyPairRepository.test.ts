import MockDate from "mockdate";
import { KeyPair } from "@lindorm-io/key-pair";
import { KeyPairRepository } from "./KeyPairRepository";
import { MOCK_MONGO_OPTIONS, MOCK_UUID } from "../../test/mocks/data";
import { MongoConnection, RepositoryEntityNotFoundError } from "@lindorm-io/mongo";
import { winston } from "../../logger";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("KeyPairRepository", () => {
  let repository: KeyPairRepository;
  let keyPair: KeyPair;

  beforeEach(async () => {
    const mongo = new MongoConnection(MOCK_MONGO_OPTIONS);
    await mongo.connect();

    repository = new KeyPairRepository({
      db: mongo.getDatabase(),
      logger: winston,
    });
    keyPair = new KeyPair({
      id: MOCK_UUID,

      algorithm: "algorithm",
      expires: new Date(),
      passphrase: "passphrase",
      privateKey: "privateKey",
      publicKey: "publicKey",
      type: "type",
    });
  });

  test("should create", async () => {
    await expect(repository.create(keyPair)).resolves.toMatchSnapshot();
  });

  test("should update", async () => {
    await repository.create(keyPair);

    await expect(repository.update(keyPair)).resolves.toMatchSnapshot();
  });

  test("should find", async () => {
    await repository.create(keyPair);

    await expect(repository.find({ type: "type" })).resolves.toMatchSnapshot();
  });

  test("should find many", async () => {
    await repository.create(keyPair);
    await repository.create(
      new KeyPair({
        algorithm: "other",
        privateKey: "privateKey",
        publicKey: "publicKey",
        type: "type",
      }),
    );

    await expect(repository.findMany({ type: "type" })).resolves.toMatchSnapshot();
  });

  test("should remove", async () => {
    await repository.create(keyPair);

    await expect(repository.remove(keyPair)).resolves.toBe(undefined);
    await expect(repository.find({ id: keyPair.id })).rejects.toStrictEqual(expect.any(RepositoryEntityNotFoundError));
  });
});
