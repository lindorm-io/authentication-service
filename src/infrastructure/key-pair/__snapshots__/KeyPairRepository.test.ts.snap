// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyPairRepository should create 1`] = `
KeyPair {
  "_algorithm": "algorithm",
  "_allowed": true,
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "key_pair_created",
      "payload": Object {
        "algorithm": "algorithm",
        "allowed": true,
        "expires": 2020-01-01T07:00:00.000Z,
        "passphrase": "passphrase",
        "privateKey": "privateKey",
        "publicKey": "publicKey",
        "type": "type",
      },
    },
  ],
  "_expires": 2020-01-01T07:00:00.000Z,
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_passphrase": "passphrase",
  "_privateKey": "privateKey",
  "_publicKey": "publicKey",
  "_type": "type",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`KeyPairRepository should find 1`] = `
KeyPair {
  "_algorithm": "algorithm",
  "_allowed": true,
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "key_pair_created",
      "payload": Object {
        "algorithm": "algorithm",
        "allowed": true,
        "expires": 2020-01-01T07:00:00.000Z,
        "passphrase": "passphrase",
        "privateKey": "privateKey",
        "publicKey": "publicKey",
        "type": "type",
      },
    },
  ],
  "_expires": 2020-01-01T07:00:00.000Z,
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_passphrase": "passphrase",
  "_privateKey": "privateKey",
  "_publicKey": "publicKey",
  "_type": "type",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`KeyPairRepository should find many 1`] = `
Array [
  KeyPair {
    "_algorithm": "algorithm",
    "_allowed": true,
    "_created": 2020-01-01T07:00:00.000Z,
    "_events": Array [
      Object {
        "date": 2020-01-01T07:00:00.000Z,
        "name": "key_pair_created",
        "payload": Object {
          "algorithm": "algorithm",
          "allowed": true,
          "expires": 2020-01-01T07:00:00.000Z,
          "passphrase": "passphrase",
          "privateKey": "privateKey",
          "publicKey": "publicKey",
          "type": "type",
        },
      },
    ],
    "_expires": 2020-01-01T07:00:00.000Z,
    "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "_passphrase": "passphrase",
    "_privateKey": "privateKey",
    "_publicKey": "publicKey",
    "_type": "type",
    "_updated": 2020-01-01T07:00:00.000Z,
    "_version": 0,
  },
  KeyPair {
    "_algorithm": "other",
    "_allowed": true,
    "_created": 2020-01-01T07:00:00.000Z,
    "_events": Array [
      Object {
        "date": 2020-01-01T07:00:00.000Z,
        "name": "key_pair_created",
        "payload": Object {
          "algorithm": "other",
          "allowed": true,
          "expires": null,
          "passphrase": null,
          "privateKey": "privateKey",
          "publicKey": "publicKey",
          "type": "type",
        },
      },
    ],
    "_expires": null,
    "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "_passphrase": null,
    "_privateKey": "privateKey",
    "_publicKey": "publicKey",
    "_type": "type",
    "_updated": 2020-01-01T07:00:00.000Z,
    "_version": 0,
  },
]
`;

exports[`KeyPairRepository should update 1`] = `
KeyPair {
  "_algorithm": "algorithm",
  "_allowed": true,
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "key_pair_created",
      "payload": Object {
        "algorithm": "algorithm",
        "allowed": true,
        "expires": 2020-01-01T07:00:00.000Z,
        "passphrase": "passphrase",
        "privateKey": "privateKey",
        "publicKey": "publicKey",
        "type": "type",
      },
    },
  ],
  "_expires": 2020-01-01T07:00:00.000Z,
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_passphrase": "passphrase",
  "_privateKey": "privateKey",
  "_publicKey": "publicKey",
  "_type": "type",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 1,
}
`;
